### PR TITLE
Groovy 8562 wrong closure delegation for property access and compile static

### PR DESCRIPTION
**Pull request by MeneDev**
Roundup:

The StaticTypeCheckingVisitor would not look consider FieldNode to potentially delegate to a different Field. I think this could only occur for name collisions, i.e. the owner has a field with the same name as the delegate.
Calling tryVariableExpressionAsProperty solved this, however this broke other tests. To solve this I cheked IMPLICIT_RECEIVER, if unset and PROPERTY_OWNER is present, checkOrMarkPrivateAccess must be called to ensure the accessors are generated.

The VariableExpressionTransformer ignored delegation metadata when a private field accessor was present. I inversed that logic to give delegation metadata precedence over private field accessors.

Build passes https://travis-ci.org/MeneDev/groovy/builds/382561549

### CODE REVIEW

**Comment by danielsun1106** [comment]
@paulk-asert It looks good for me. What do you think about the PR?

**Comment by paulk-asert** [comment]
on my TODO list to review - at first glance I am optimistic

File: src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java [hunk 1]
```diff
@@ -596,6 +597,26 @@ public void visitVariableExpression(VariableExpression vexp) {
...
+                if (tryVariableExpressionAsProperty(vexp, vexp.getName())) {
+                    // IMPLICIT_RECEIVER is handled elsewhere
+                    // however other access needs to be fixed for private access
+                    if (vexp.getNodeMetaData(StaticTypesMarker.IMPLICIT_RECEIVER) == null) {
+                        boolean lhsOfEnclosingAssignment = isLHSOfEnclosingAssignment(vexp);
+                        ClassNode owner = (ClassNode) vexp.getNodeMetaData(StaticCompilationMetadataKeys.PROPERTY_OWNER);
+                        if (owner != null) {
+                            fieldNode = owner.getField(vexp.getName());
+                            vexp.setAccessedVariable(fieldNode);
+                            checkOrMarkPrivateAccess(vexp, fieldNode, lhsOfEnclosingAssignment);
```
Comment (danielsun1106) [inline src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java]: `lhsOfEnclosingAssignment` is only used here(in the `if` block)

**Comment by paulk-asert** [comment]
Merged. Thanks!

**Pull request closed without merging**